### PR TITLE
Lighter weight matches

### LINKED ISSUE #121322: Webviews should use onWillMoveEditor
Currently Webviews (and by extension custom editors) require the input to know what group they're being opened in to handle the moving of the webview without a re-render. I recently added `onWillMoveEditor` for notebooks to handle this case. This should be adopted to remove the dependency on knowing about the group.

### PR DESCRIPTION
**Pull request by lramos15**
This removes the `matchesEditor` call from the `editorGroupFinder` as having a separate implementation led to a lot of bugs especially when new editor types such as the `MergeEditor` were forgotten. Given that the `MergeEditor` (and most likely future editors) adopt a contrib model I believe the logic should be on them to decide whether or not a typed / untyped editor properly matches the current instanceof of that EditorInput.

This PR adds a more lightweight fallback to all matches calls when an `override` is not present on the `UntypedEditorInput`. This is normally the case when we are resolving the `UntypedEditorInput` due to some technical debt requiring the group to be known prior to resolution. 

While this is not without it's own bugs, such as an untyped `test.ipynb` matching a text editor open with `test.ipynb` resource and a notebook open with a `test.ipynb` I feel that this is acceptible match result when the editor id of this unresolved input is ambiguous. Another downside is this can be considered "spreading the debt" to each input rather than a centralized location, but there might be more places where we have bugs due to matches failing. This weaker match can be removed if we ever resolve #121322

### CODE REVIEW

File: src/vs/workbench/common/editor/editorInput.ts [hunk 1]
```diff
@@ -262,12 +262,9 @@ export abstract class EditorInput extends AbstractEditorInput {
...
 		const otherInputEditorId = otherInput.options?.override;
 
-		if (this.editorId === undefined) {
-			return false; // untyped inputs can only match for editors that have adopted `editorId`
-		}
-
-		if (this.editorId !== otherInputEditorId) {
-			return false; // untyped input uses another `editorId`
+		// If the overrides are both defined and don't match that means they're separate inputs
+		if (this.editorId !== otherInputEditorId && otherInputEditorId !== undefined) {
```
Comment (bpasero) [inline src/vs/workbench/common/editor/editorInput.ts:266]: Are we missing the case of `this.editorId === undefined`?

File: src/vs/workbench/contrib/customEditor/browser/customEditorInput.ts [hunk 2]
```diff
@@ -244,7 +244,7 @@ export class CustomEditorInput extends LazilyResolvedWebviewEditorInput {
 			return true;
 		}
 		return this === other || (other instanceof CustomEditorInput
-			&& this.viewType === other.viewType
+			&& (this.viewType === other.viewType || other.viewType === undefined)
```
Comment (bpasero) [inline src/vs/workbench/contrib/customEditor/browser/customEditorInput.ts:247]: Shouldn't we apply this only to matching on untyped inputs? How can a typed input now have a view type here? I think here we now match somewhat differently from the `MergeEditorInput`.

**Review by bpasero — CHANGES_REQUESTED** [review]
Besides my feedback, I think we should also search for all implementations of `matches` and find those that have not adopted the editor resolver API and clean them up to not match on untyped editors but only typed editors. For example:

https://github.com/microsoft/vscode/blob/b52b57134d7e1d3fa0c76b9774ce6ef727b2fd44/src/vs/workbench/contrib/extensions/common/runtimeExtensionsInput.ts#L41-L46

This imho should not call to `super.matches` should it? In general I feel those `super.matches` calls are less than ideal.

File: src/vs/workbench/contrib/customEditor/browser/customEditorInput.ts (hunk 2 above)
Comment (lramos15) [reply src/vs/workbench/contrib/customEditor/browser/customEditorInput.ts]: Sorry you're correct, `super.matches` covers the untyped use case here

File: src/vs/workbench/common/editor/editorInput.ts (hunk 1 above)
Comment (lramos15) [reply src/vs/workbench/common/editor/editorInput.ts]: Yep, I forgot `this.editorId` could even be undefined.

**Comment by lramos15** [comment]
> I think we should also search for all implementations of `matches` and find those that have not adopted the editor resolver API and clean them up to not match on untyped editors but only typed editors. For example:

I agree, we can do this in a separate PR though as they're unrelated

**Review by bpasero — APPROVED** [review]

**Comment by bpasero** [comment]
Yes another PR is fine.

**Pull request merged**